Horizon/Horizon/Sources/Features/Assist/AssistChat/Scratch/Goal/SelectCourseGoal.swift

```diff
@@ -0,0 +1,98 @@
+//
+// This file is part of Canvas.
+// Copyright (C) 2025-present  Instructure, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+//
+
+import Combine
+import Core
+import Foundation
+
+class SelectCourseGoal: Goal {
+    private let cedar: DomainService
+    private let environment: AssistDataEnvironment
+    private let userID: String
+
+    init(
+        environment: AssistDataEnvironment,
+        cedar: DomainService = DomainService(.cedar),
+        userID: String = AppEnvironment.shared.currentSession?.userID ?? ""
+    ) {
+        self.environment = environment
+        self.cedar = cedar
+        self.userID = userID
+    }
+
+    override
+    func isRequested() -> Bool {
+        environment.courseID.value == nil
+    }
+
+    override
+    func execute(response: String? = nil, history: [AssistChatMessage] = []) -> AnyPublisher<AssistChatMessage?, any Error> {
+        guard let response = response, response.isNotEmpty else {
+            return initialPrompt(history: history)
+        }
+
+        return selectCourseFrom(response: response, history: history)
+    }
+
+    private func selectCourseFrom(response: String, history: [AssistChatMessage]) -> AnyPublisher<AssistChatMessage?, any Error> {
+        courses.flatMap { [weak self] courseOptions in
+            guard let self = self else {
+                return Just<AssistChatMessage?>(nil).setFailureType(to: Error.self).eraseToAnyPublisher()
+            }
+            let courseNames = courseOptions.compactMap { $0.course.name }
+            return self.choose(from: courseNames, with: response, using: self.cedar)
+                .map { courseSelected in
+                    if  let courseSelected = courseSelected,
+                        let courseID = courseOptions.first(where: { courseSelected.contains($0.course.name ?? "") == true })?.courseID {
+                        self.environment.courseID.accept(courseID)
```

Comment:
Since this is a different closure compared to where self was captured weakly, it can cause leaks. One solution could be that we define the `weak var self = self` at the very start of `selectCourseFrom` so it's local the function.